### PR TITLE
8274243: Implement fast-path for ASCII-compatible CharsetEncoders on aarch64

### PR DESCRIPTION
**Pull request by ghost**
Implementation of ISO/ASCII char set encoding, extending current implementation with ASCII encoding support.

The motivation is found in the original x86 issue ([JDK-8274242](https://bugs.openjdk.java.net/browse/JDK-8274242)).

Implementation with slight focus on balance between footprint and efficiency, trying to utilise a dual SIMD path (e.g. Neoverse N1) for the additional Ascii-check and avoid performance loss in the ISO-only case.

- Interleaved ISO and ASCII check code.
- Avoid 'umaxv' in the ISO main flow.
- Using post inc in main loop.
- Retain 8-char loop.
- Removing conditional prefetch (no upside).
- Adding ISO-8859-1 to encode-decode benchmark.

Testing: tier1-6

The revised version compares like this (master vs. update).
```
Benchmark                   (size)       (type)  Mode  Cnt    Score   Error  Units
CharsetEncodeDecode.encode   16384        UTF-8  avgt   30   17.920 ± 0.229  us/op
CharsetEncodeDecode.encode   16384         BIG5  avgt   30   18.867 ± 0.356  us/op
CharsetEncodeDecode.encode   16384  ISO-8859-15  avgt   30   17.419 ± 0.220  us/op
CharsetEncodeDecode.encode   16384   ISO-8859-1  avgt   30    6.200 ± 0.134  us/op
CharsetEncodeDecode.encode   16384        ASCII  avgt   30   17.149 ± 0.219  us/op
CharsetEncodeDecode.encode   16384       UTF-16  avgt   30  135.115 ± 1.440  us/op
```
```
Benchmark                   (size)       (type)  Mode  Cnt    Score   Error  Units
CharsetEncodeDecode.encode   16384        UTF-8  avgt   30    9.018 ± 0.179  us/op
CharsetEncodeDecode.encode   16384         BIG5  avgt   30   10.550 ± 0.470  us/op
CharsetEncodeDecode.encode   16384  ISO-8859-15  avgt   30    8.843 ± 0.187  us/op
CharsetEncodeDecode.encode   16384   ISO-8859-1  avgt   30    6.406 ± 0.155  us/op
CharsetEncodeDecode.encode   16384        ASCII  avgt   30    8.822 ± 0.173  us/op
CharsetEncodeDecode.encode   16384       UTF-16  avgt   30  135.195 ± 1.432  us/op
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8274243](https://bugs.openjdk.java.net/browse/JDK-8274243): Implement fast-path for ASCII-compatible CharsetEncoders on aarch64

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk18 pull/20/head:pull/20` \
`$ git checkout pull/20`

Update a local copy of the PR: \
`$ git checkout pull/20` \
`$ git pull https://git.openjdk.java.net/jdk18 pull/20/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20`

View PR using the GUI difftool: \
`$ git pr show -t 20`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk18/pull/20.diff">https://git.openjdk.java.net/jdk18/pull/20.diff</a>

</details>

### CODE REVIEW

**Comment by ghost** [comment]
Benchmarks, master vs. update (ran on Aurora/Ampere Altra):

```
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16384-type:ASCII ........77.55%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16384-type:BIG5 .........76.71%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16384-type:ISO_8859_1 ...-2.31%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16384-type:ISO_8859_15 ..75.58%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16384-type:UTF_16 ....... 1.04%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16384-type:UTF_8 ........76.90%
```
Note that ISO-8859-1 compares with the old intrinsic implementation (essentially the same) and that UTF-16 does not utilise the intrinsic.

Runs that show the more pessimistic speed-up, when processing 2^n - 1 chars.
```
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:2047-type:ASCII .........72.97%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:2047-type:BIG5 ..........64.46%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:2047-type:ISO_8859_1 ....-1.67%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:2047-type:ISO_8859_15 ...70.85%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:2047-type:UTF_16 ........-4.60%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:2047-type:UTF_8 .........70.44%

openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:511-type:ASCII ..........60.35%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:511-type:BIG5 ...........52.61%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:511-type:ISO_8859_1 ..... 1.75%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:511-type:ISO_8859_15 ....61.45%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:511-type:UTF_16 .........-1.01%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:511-type:UTF_8 ..........59.46%

openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:255-type:ASCII ..........54.26%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:255-type:BIG5 ...........42.82%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:255-type:ISO_8859_1 .....-0.54%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:255-type:ISO_8859_15 ....64.86%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:255-type:UTF_16 .........-0.09%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:255-type:UTF_8 ..........60.44%

openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:127-type:ASCII ..........51.51%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:127-type:BIG5 ...........46.54%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:127-type:ISO_8859_1 .....-0.32%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:127-type:ISO_8859_15 ....56.48%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:127-type:UTF_16 ......... 0.44%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:127-type:UTF_8 ..........54.84%
```
Runs to illustrate the threshold effect between the loops in the implementation.
```
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:32-type:ASCII ...........32.30%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:32-type:BIG5 ............31.93%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:32-type:ISO_8859_1 ......-0.02%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:32-type:ISO_8859_15 .....37.92%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:32-type:UTF_16 .......... 4.45%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:32-type:UTF_8 ...........40.35%

openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:31-type:ASCII ...........20.06%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:31-type:BIG5 ............21.64%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:31-type:ISO_8859_1 ......-1.13%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:31-type:ISO_8859_15 .....27.04%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:31-type:UTF_16 .......... 1.20%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:31-type:UTF_8 ...........24.72%

openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16-type:ASCII ...........19.37%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16-type:BIG5 ............20.20%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16-type:ISO_8859_1 ......-1.01%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16-type:ISO_8859_15 .....29.16%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16-type:UTF_16 .......... 0.34%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:16-type:UTF_8 ...........25.35%

openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:15-type:ASCII ...........13.03%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:15-type:BIG5 ............13.74%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:15-type:ISO_8859_1 ......-0.13%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:15-type:ISO_8859_15 .....19.26%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:15-type:UTF_16 .......... 0.78%
openjdk.bench.java.nio.CharsetEncodeDecode.encode-size:15-type:UTF_8 ...........17.70%
```

**Comment by ghost** [comment]
Using the microbenchmarks provided by @carterkozak here: https://github.com/carterkozak/stringbuilder-encoding-performance, comparing master vs. update as follows:
```
Benchmark                                                      (charsetName)                          (message)  (timesToAppend)  Mode  Cnt    Score    Error  Units
EncoderBenchmarks.charsetEncoder                                       UTF-8     This is a simple ASCII message                3  avgt    4  151.025 ± 28.111  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  This is a message with unicode 😊                3  avgt    4  323.254 ±  5.648  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8     This is a simple ASCII message                3  avgt    4  244.375 ± 98.844  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  This is a message with unicode 😊                3  avgt    4  405.415 ±  5.947  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8     This is a simple ASCII message                3  avgt    4  728.172 ± 22.419  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  This is a message with unicode 😊                3  avgt    4  859.015 ± 90.541  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8     This is a simple ASCII message                3  avgt    4  117.044 ± 11.484  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  This is a message with unicode 😊                3  avgt    4  483.399 ± 38.614  ns/op
```

```
Benchmark                                                      (charsetName)                          (message)  (timesToAppend)  Mode  Cnt    Score    Error  Units
EncoderBenchmarks.charsetEncoder                                       UTF-8     This is a simple ASCII message                3  avgt    4  113.954 ±  7.657  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  This is a message with unicode 😊                3  avgt    4  353.266 ± 10.124  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8     This is a simple ASCII message                3  avgt    4  196.643 ± 52.954  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  This is a message with unicode 😊                3  avgt    4  429.157 ± 11.506  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8     This is a simple ASCII message                3  avgt    4  728.138 ± 34.898  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  This is a message with unicode 😊                3  avgt    4  859.697 ± 61.397  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8     This is a simple ASCII message                3  avgt    4  117.269 ±  6.623  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  This is a message with unicode 😊                3  avgt    4  491.559 ± 68.169  ns/op
```
Note: The above was ran on a local dev-machine typically producing less than _perfectly_ consistent results.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -4923,111 +4923,118 @@ void MacroAssembler::fill_words(Register base, Register cnt, Register value)
...
-    BIND(DONE);
-}
+  Register cnt = res;
+  Register max = rscratch1;
+  Register chk = rscratch2;
+
+  prfm(Address(src), PLDL1STRM);
+  movw(cnt, len);
+
+#define ASCII(insn) if (ascii) { insn; }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:4956]:
```suggestion
#define ASCII(insn) do { if (ascii) { insn; } } while (0)
```
If a macro is to be used here, it should be an else-safe one, as a matter of good practice.

**Comment by theRealAph** [comment]
I don't think this should go straight into the 18 release branch. It looks OK for mainline.

**Comment by ghost** [comment]
> I don't think this should go straight into the 18 release branch. It looks OK for mainline.

Any particular reason it should not be included in JDK-18?

**Comment by theRealAph** [comment]
> > I don't think this should go straight into the 18 release branch. It looks OK for mainline.
> 
> Any particular reason it should not be included in JDK-18?

We're in RDP1 since December 9. This means that any enhancement is covered by the Late-Enhancement Request Process.
https://openjdk.java.net/jeps/3#Late-Enhancement-Request-Process . I don't think this patch is urgent enough for that.

**Comment by ghost** [comment]
> We're in RDP1 since December 9. This means that any enhancement is covered by the Late-Enhancement Request Process. https://openjdk.java.net/jeps/3#Late-Enhancement-Request-Process . I don't think this patch is urgent enough for that.

It has been classified as a performance regression (bug) in line with the x86 issue (JDK-8274242). Do you mean we should change this _now_? Aarch64 would be the only platform not to address the issue in JDK-18.

**Comment by theRealAph** [comment]
> > We're in RDP1 since December 9. This means that any enhancement is covered by the Late-Enhancement Request Process. https://openjdk.java.net/jeps/3#Late-Enhancement-Request-Process . I don't think this patch is urgent enough for that.
> 
> It has been classified as a performance regression (bug) in line with the x86 issue (JDK-8274242). Do you mean we should change this _now_? Aarch64 would be the only platform not to address the issue in JDK-18.

I see your point, and that is a significant difference. It's unfortunate that AArch64 got this patch late in the process, but as it's a bug, and Java performance isn't supposed to regress in a release, it does make sense to fix it. Having said that, I've had a lot of bad experiences with late patches, and it'd be a dreadful shame if we broke AArch64 string handling.

I suppose the best way to proceed with this is to have two AArch64 reviewers go through the patch instruction by instruction, just to make sure the test covers all corner cases. Thanks.

**Comment by RogerRiggs** [comment]
Its viable to commit to the main line and allow it to have some bake time before requesting it to be backported.
That would allow some time to build confidence about the change; it might not make the first JDK 18 release but would come in later.
There is no problem requesting approval for the change but it should go through that process.

**Comment by ghost** [comment]
So this will obviously be prolonged. Closing and moving to 19.

**Comment by theRealAph** [comment]
> So this will obviously be prolonged. Closing and moving to 19.

That sounds right. Let's get it into mainline soon, and we can do a backport to 18.1.
